### DISCUSSION TITLE
I want to write my own NTRIP client

### DISCUSSION THREAD

**Discussion by WH2425** [semuconsulting/pyrtcm#62] [Q&A] [answered]
Hello, I hope to obtain the last incomplete data when the stream terminates .Because I can only use io.BytesIO(data) to simulate the stream for real-time decoding.I would like to be able to return unused bytes and None at the end of the stream.

```
if 0 < len(data) < size:  # truncated stream
    raise RTCMStreamError(
        "Serial stream terminated unexpectedly. "
        f"{size} bytes requested, {len(data)} bytes returned."
    )
```

```
for data in ntrip.record_data_yield():
    byte_stream = io.BytesIO(data)
    rtr = RTCMReader(byte_stream)
    for raw_data, parsed_data in rtr:
        if parsed_data == None :
            pass
            # I hope to obtain the last incomplete data when the stream terminates here.
```

**Comment by semuadmin** [comment]
Hi @WH2425 

Some additional context would be helpful here...

If your intention here is to parse a *live* NTRIP data stream, there may be a simpler approach...

`pyrtcm.RTCMReader` is able to parse an NTRIP TCP Socket stream directly via it's underlying [`SocketWrapper`](https://github.com/semuconsulting/pyrtcm/blob/main/src/pyrtcm/socket_wrapper.py) class - no need for a `BytesIO` intermediary. See the README for details.

if you require more advanced NTRIP Client functionality, including support for encrypted (TLS) and encoded (chunked, gzip, etc.) NTRIP data streams and automatic connection retries, have a look at the [gnssntripclient.GNSSNTRIPClient](https://github.com/semuconsulting/pygnssutils/blob/main/src/pygnssutils/gnssntripclient.py) class in [pygnssutils](https://github.com/semuconsulting/pygnssutils), which implements a more advanced version of [`SocketWrapper`](https://github.com/semuconsulting/pygnssutils/blob/main/src/pygnssutils/socketwrapper.py). There's some basic example code [here](https://github.com/semuconsulting/pyrtcm/blob/main/examples/rtcm_ntrip_client.py).

If, on the other hand, you're parsing an *offline* NTRIP data log, then you could simply concatenate the individual `data` segments from your `ntrip.record_data_yield()` iteration into a single contiguous binary data file, and then parse that file using `pyrtcm.RTCMReader` in the normal manner i.e. something along these lines...

```python
with open("ntripdata.bin", 'wb') as ntripfile:
    for data in ntrip.record_data_yield():
        ntripfile.write(data)

with open("ntripdata.bin", 'rb') as stream:
    rtr = RTCMReader(stream)
    for raw_data, parsed_data in rtr:
        if parsed_data is not None:
            print(parsed_data)

```

**Reply by WH2425** [reply]
Thank you so much，I'm trying to parse a NTRIP TCP socket stream directly through its underlying SocketWrapper class and migrate the functionality I need under a 'for raw_data, parsed_data in rtr:' loop.pyrtcm.But RTCMReader does not seem to have a reconnection mechanism, and the connection may be disconnected midway

**Reply by semuadmin** [reply]
>.But RTCMReader does not seem to have a reconnection mechanism,...

No, because it's just an RTCM parser, not a full-blown NTRIP client. If you're after NTRIP client functionality (*and it sounds like you are*), look to `pygnssutils.gnssntripclient` (*which does have connection retry capabilities*) e.g. the example I provided above.

**Reply by WH2425** [reply]
Okay, thank you very much. I am trying to implement NTRIP client functionality myself.
```python
stream = ntrip.client
rtr = RTCMReader(stream)
for raw_data, parsed_data in rtr:
	rtcm_id = parsed_data.DF002
```

 I would like to ask again when this loop will exit, whether it will be a timeout when requesting the data stream or something else. I'm going to reconnect and start RTCMReader again after it exits the loop

**Reply by WH2425** [reply]
I plan to re-establish the NTRIP connection and restart the RTCMReader after the RTCMReader exits the loop.Do you think this is reasonable?

**Reply by semuadmin** [reply, marked as answer]
> I plan to re-establish the NTRIP connection and restart the RTCMReader after the RTCMReader exits the loop.Do you think this is reasonable?

Your code, your call.

You might, however, want to familiarise yourself with the NTRIP Protocol specification and Client Best Practice guidelines, which have very specific provisions on connection retries. Failure to comply with those provisions is likely to get you banned by NTRIP service providers.

[NTRIP Protocols Version 1 & Version 2](https://rtcm.myshopify.com/collections/differential-global-navigation-satellite-dgnss-standards/products/copy-of-differential-gnss-package-both-of-the-current-standards-10410-1-and-10410-2)

[RTCM Paper 2023-SC104-1344, NTRIP Client Devices / Best Practices](https://rtcm.myshopify.com/collections/differential-global-navigation-satellite-dgnss-standards/products/rtcm-paper-2023-sc104-1344-ntrip-client-devices-best-practices)

Bear in mind that you don't *necessarily* need to use `pyrtcm` at all. If your intention is simply to send RTK corrections to a compatible GNSS receiver (*like for example the u-blox ZED-F9P*) then you can simply pass the NTRIP binary data stream (`'content-type': 'gnss/data'`) directly to the device. The only value in using `pyrtcm` in real time is if you actually want to see what's inside each RTCM message as it arrives.

**Comment by WH2425** [comment]
Hi @semuadmin 
I have some doubts I would like to ask you, as follows:

1. What will cause the rtcm parser to exit?

When I exit the rtcm parser, I check if the TCP connection is not interrupted. Sometimes the TCP connection is not interrupted. In addition, I found that the exit of the rtcm parser is not necessarily caused by a timeout in getting data from TCP.

2. I also want to know whether I need to re-initialize when re-entering the rtcm parser?Currently, I detect that the TCP connection is normal and do not perform initialization, but directly enter the rtcm parser again. If the TCP connection is disconnected, I will reconnect first, then perform initialization and then enter the rtcm parser.

```python
stream = ntrip.client
rtr = RTCMReader(stream)
```

I am looking forward to your answer, thank you very much
